### ISSUE TITLE
Benchmark timeout when benchtime set too long

### ISSUE THREAD

**Issue by lkxiaolou** [golang/go#49973]
### What version of Go are you using (`go version`)?

<pre>
go1.15
</pre>

### Does this issue reproduce with the latest release?
yes

### What did you do?

When benchmark function execute time is very short, set benchtime=5s / 10s / 20s is ok.
```
BenchmarkNewObject-2            1000000000               0.279 ns/op
```
But set benchtime=120s will execute time out.

```
*** Test killed with quit: ran too long (11m0s).
```

### What did you expect to see?
Do not timeout

### What did you see instead?
```
*** Test killed with quit: ran too long (11m0s).
```

**Comment by seankhliao** [comment]
Duplicate of #12446

**Comment by lkxiaolou** [comment]
> Duplicate of #12446

Not the same,  see #49974,  int64 overflow, may be a bug

**Comment by mvdan** [comment]
I agree that this might not be a duplicate - the code doesn't seem to be safe against overflows.

That said, the fix doesn't seem complete. Not all overflows will result in negative numbers.

**Comment by lkxiaolou** [comment]
> I agree that this might not be a duplicate - the code doesn't seem to be safe against overflows.
> 
> That said, the fix doesn't seem complete. Not all overflows will result in negative numbers.

Sorry, Is there any other cases?

**Comment by mvdan** [comment]
https://go.dev/play/p/MBFpdZOOSgx

That example also has some better overflow detection that you can enable and try for yourself.

Note that there are multiple operations involved in testing/benchmark.go, including multiplications and additions. Proper overflow checking won't be as trivial as in my example.

**Comment by mvdan** [comment]
Also, a fix would need to come with a test :)

**Comment by lkxiaolou** [comment]
> Also, a fix would need to come with a test :)

Done, see new PR #50023
